x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/HasApplicationPrivilegesIT.java

```diff
@@ -0,0 +1,215 @@
+import org.elasticsearch.test.TestSecurityClient;
+import org.elasticsearch.test.XContentTestUtils;
+import org.elasticsearch.xcontent.ObjectPath;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.core.security.authc.support.UsernamePasswordToken;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.permission.ResourcePrivileges;
+import org.elasticsearch.xpack.core.security.authz.privilege.ConfigurableClusterPrivilege;
+import org.elasticsearch.xpack.core.security.user.User;
+import org.hamcrest.Matchers;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.hamcrest.Matchers.aMapWithSize;
+import static org.hamcrest.Matchers.equalTo;
+
+public class HasApplicationPrivilegesIT extends SecurityInBasicRestTestCase {
+
+    private TestSecurityClient securityClient;
+
+    @Before
+    public void setupClient() {
+        securityClient = new TestSecurityClient(adminClient());
+    }
+
+    public void testUserWithWildcardPrivileges() throws Exception {
+        var appName = randomApplicationName();
+
+        final Set<String> allApplications = new HashSet<>();
+        CheckedBiConsumer<String, Integer, IOException> createApplicationPrivilege = (app, index) -> {
+            createApplicationPrivilege(app, "priv_" + index, randomArray(1, 4, String[]::new, () -> randomActionName()));
+            allApplications.add(appName);
+        };
+
+        for (int i = randomIntBetween(0, 2); i > 0; i--) {
+            // Create 0 or more application privileges for this application
+            createApplicationPrivilege.accept(appName, i);
+        }
+
+        for (int i = randomIntBetween(0, 3); i > 0; i--) {
+            // Create 0 or more application privileges for other applications
+            createApplicationPrivilege.accept(randomValueOtherThan(appName, this::randomApplicationName), i);
+        }
+
+        // Define a role with all privileges (by wildcard) for this application
+        var roleName = randomAlphaOfLengthBetween(6, 10);
+        var singleAppOnly = randomBoolean();
+        createRole(roleName, singleAppOnly ? appName : "*", new String[] { "*" }, new String[] { "*" });
+
+        final Set<String> allRoles = new HashSet<>();
+        allRoles.add(roleName);
+
+        for (int i = randomIntBetween(0, 3); i > 0; i--) {
+            // Create 0 or more additional roles with privileges for one of the applications
+            var extraRoleName = randomValueOtherThanMany(allRoles::contains, () -> randomAlphaOfLengthBetween(8, 16));
+            createRole(
+                extraRoleName,
+                randomFrom(allApplications),
+                new String[] { "priv_" + i },
```

Comment:
Yes, I care in the positive sense.
My intent was that this test would, across enough randomised runs, generate a great variety of scenarios that would not otherwise be tested.

That includes the user being granted a privilege that hasn't been defined. An undefined privilege should not affect the result of `has_privileges` for a user with `*`,`*`,`*`

I'll see if I can add some more comments in to explain some of the behaviours.